src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,690 @@
+		{
+#if !FEATURE_ILLINK
+			if (IsExcluded (nav))
+				return;
+#endif
+			ProcessTypeChildren (type, nav);
+
+			if (!type.HasNestedTypes)
+				return;
+
+			foreach (TypeDefinition nested in type.NestedTypes) {
+				var iterator = nav.SelectChildren ("type", _ns);
+				while (iterator.MoveNext ()) {
+					if (nested.Name == GetAttribute (iterator.Current, "name"))
+						ProcessTypeChildren (nested, iterator.Current);
+				}
+			}
+		}
+
+		void ProcessTypeChildren (TypeDefinition type, XPathNavigator nav)
+		{
+			if (nav.HasChildren) {
+				ProcessSelectedFields (nav, type);
+				ProcessSelectedMethods (nav, type);
+				ProcessSelectedProperties (nav, type);
+			}
+		}
+
+		void ProcessSelectedFields (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator fields = nav.SelectChildren ("field", _ns);
+			if (fields.Count == 0)
+				return;
+
+			ProcessFields (type, fields);
+		}
+
+		void ProcessSelectedMethods (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator methods = nav.SelectChildren ("method", _ns);
+			if (methods.Count == 0)
+				return;
+
+			ProcessMethods (type, methods);
+		}
+
+		void ProcessSelectedProperties (XPathNavigator nav, TypeDefinition type)
+		{
+			XPathNodeIterator properties = nav.SelectChildren ("property", _ns);
+			if (properties.Count == 0)
+				return;
+
+			ProcessProperties (type, properties);
+		}
+
+		static TypePreserve GetTypePreserve (XPathNavigator nav)
+		{
+			string attribute = GetAttribute (nav, _preserve);
+			if (string.IsNullOrEmpty (attribute))
+				return nav.HasChildren ? TypePreserve.Nothing : TypePreserve.All;
+
+			if (Enum.TryParse (attribute, true, out TypePreserve result))
+				return result;
+			return TypePreserve.Nothing;
+		}
+
+		void ProcessFields (TypeDefinition type, XPathNodeIterator iterator)
+		{
+			while (iterator.MoveNext ())
+				ProcessField (type, iterator);
+		}
+
+		protected virtual void ProcessField (TypeDefinition type, XPathNodeIterator iterator)
+		{
+#if !FEATURE_ILLINK
+			if (IsExcluded (iterator.Current))
+				return;
+#endif
+			string value = GetSignature (iterator.Current);
+			if (!String.IsNullOrEmpty (value))
+				ProcessFieldSignature (type, value, iterator);
+
+			value = GetAttribute (iterator.Current, "name");
+			if (!String.IsNullOrEmpty (value))
+				ProcessFieldName (type, value, iterator);
+		}
+
+		void ProcessFieldSignature (TypeDefinition type, string signature, XPathNodeIterator iterator)
+		{
+			FieldDefinition field = GetField (type, signature);
+			if (field != null) {
+				ArrayBuilder<Attribute> attributes = ProcessAttributes (iterator.Current.SelectChildren ("attribute", _ns));
+				DynamicallyAccessedMemberTypes fieldAnnotation = ParseKinds (attributes);
+				if (fieldAnnotation != DynamicallyAccessedMemberTypes.None)
+					_fields[field] = fieldAnnotation;
+			}
+		}
+
+		void ProcessFieldName (TypeDefinition type, string name, XPathNodeIterator iterator)
```

Comment:
Again - why separate method - this makes the code much harder to read - typically if something is extracted into a method one expects it to be called from multiple places. These are not.